Initial commit for discover-packet

jasmingacic opened the pull request:
Dependencies are not locked but I wasn't sure which dependency tool you were using. If someone can help me out with that.
Also is there anything else missing here?

pearkes commented:
Hey there, this is looking like it is headed in the right direction, just have a few high level bits of feedback:

- Your vendoring should only include packages for Packet in this case. See [AWS](https://github.com/hashicorp/go-discover/blob/master/provider/aws/vendor/vendor.json) for an example. You did use the right tool here (govendor).
- It looks like you hardcode the test project. If possible, make this an environment variable so we can test in CI. We run automated tests for a lot of the providers, see an example with Google: https://github.com/hashicorp/go-discover/tree/master/test/tf/gce. If you can write some Terraform there we can set that up and make sure that the go-discover provider works end-to-end.
- Is a "project" in Packet similar to a "tag" in most other infrastructure providers? If you have the notion of "tags" it may be more suitable to use that.

If we can take care of those items I think this will otherwise be good!

jasmingacic commented:
@pearkes I've just fixed dependencies and added project as environment variable. 

Listing Packet devices can be done only by querying `/projects/{id}/devices` I could fetch all projects and then iterate through each to fetch devices if that's what you would like to have here.
Also there is a parameter `with_flags` but it is not yet publicly accessible.

p.s. I will be adding TF files later tomorrow.

jasmingacic commented:
@pearkes I've added TF tests.

pearkes commented on provider/packet/packet_discover_test.go:
```diff
@@ -0,0 +1,36 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(addrs) == 0 {
```
Usually like to check the explicit number setup in the TF here (2 I think).

pearkes commented on test/tf/packet/main.tf:
```diff
@@ -0,0 +1,19 @@
+provider "packet" {}
+
+resource "packet_project" "project" {
+  name = "go-discover:packet"
```
Can you add a bit of randomness here? See vsphere for [an example](https://github.com/hashicorp/go-discover/blob/master/test/tf/vsphere/main.tf#L89-L98). If this is a unique name in Packet that will help the tests from conflicting with each other.

pearkes commented on test/tf/packet/main.tf:
```diff
@@ -0,0 +1,19 @@
+
+resource "packet_device" "discover-packet" {
+  count            = 2
+  hostname         = "tf.discover"
```
Randomness here if hostnames are unique (not sure).

jasmingacic replied:
Hostnames don't have to be unique, but i will randomize it anyways.

jasmingacic commented on provider/packet/packet_discover.go:
```diff
@@ -0,0 +1,94 @@
+		addressType = "private_v4"
+	}
+
+	if addressType == "" {
```
i will remove it. It simply doesn't make sense to have it.

jasmingacic commented on provider/packet/packet_discover_test.go:
```diff
@@ -0,0 +1,42 @@
+var _ discover.Provider = (*packet.Provider)(nil)
+var _ discover.ProviderWithUserAgent = (*packet.Provider)(nil)
+
+func TestAddrs(t *testing.T) {
```
on it I will add tests for each existing type

pearkes commented:
All looks good to me! This will be in a future version of Consul.

It'd be great to add docs here as well:

https://github.com/hashicorp/consul/blob/master/website/source/docs/agent/cloud-auto-join.html.md
